### PR TITLE
Pack multiple vertex and index arrays together into growable buffers.

### PR DESCRIPTION
**Pull request by pcwalton**
This commit uses the [`offset-allocator`] crate to combine vertex and index arrays from different meshes into single buffers. Since the primary source of `wgpu` overhead is from validation and synchronization when switching buffers, this significantly improves Bevy's rendering performance on many scenes.

This patch is a more flexible version of #13218, which also used slabs. Unlike #13218, which used slabs of a fixed size, this commit implements slabs that start small and can grow. In addition to reducing memory usage, supporting slab growth reduces the number of vertex and index buffer switches that need to happen during rendering, leading to improved performance. To prevent pathological fragmentation behavior, slabs are capped to a maximum size, and mesh arrays that are too large get their own dedicated slabs.

As an additional improvement over #13218, this commit allows the application to customize all allocator heuristics. The `MeshAllocatorSettings` resource contains values that adjust the minimum and maximum slab sizes, the cutoff point at which meshes get their own dedicated slabs, and the rate at which slabs grow. Hopefully-sensible defaults have been chosen for each value.

Unfortunately, WebGL 2 doesn't support the *base vertex* feature, which is necessary to pack vertex arrays from different meshes into the same buffer. `wgpu` represents this restriction as the downlevel flag `BASE_VERTEX`. This patch detects that bit and ensures that all vertex buffers get dedicated slabs on that platform. Even on WebGL 2, though, we can combine all *index* arrays into single buffers to reduce buffer changes, and we do so.

The following measurements are on Bistro:

Overall frame time improves from 8.74 ms to 5.53 ms (1.58x speedup):
Screenshot 2024-07-09 163521

Render system time improves from 6.57 ms to 3.54 ms (1.86x speedup):
Screenshot 2024-07-09 163559

Opaque pass time improves from 4.64 ms to 2.33 ms (1.99x speedup):
Screenshot 2024-07-09 163536

### Changed

* Vertex and index buffers for meshes may now be packed alongside other buffers, for performance.
* `GpuMesh` has been renamed to `RenderMesh`, to reflect the fact that it no longer directly stores handles to GPU objects.
* Because meshes no longer have their own vertex and index buffers, the responsibility for the buffers has moved from `GpuMesh` (now called `RenderMesh`) to the `MeshAllocator` resource. To access the vertex data for a mesh, use `MeshAllocator::mesh_vertex_slice`. To access the index data for a mesh, use `MeshAllocator::mesh_index_slice`.

[`offset-allocator`]: https://github.com/pcwalton/offset-allocator

### CODE REVIEW

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:36]:
```suggestion
/// A plugin that provides GPU buffer management for uploading [`Mesh`] data.
```

Something like this, maybe. Still not quite happy with it though.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:39]:
```suggestion
/// Manages the assignment of mesh data to GPU buffers.
```

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:92]: MB or MiB?

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:118]: Why not 2?

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:160]:
```suggestion
    /// The [`Allocator`] that manages the objects in this slab.
```

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:219]:
```suggestion
/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`]. When we
```

Specifying 4 makes reading the sentence more confusing that not, imo.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:261]: HashMap seems excessive here and in SlabToReallocate. Do we have some kind of SmallVec-like map type?

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:272]: Can you move this up to under the definition of MeshAllocatorSettings ?

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:305]: Why before here, and not after? This stands out as very weird to me.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:521]:
```suggestion
    fn copy_element_data(
```

To match the terminology used elsewhere.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:1007]: You can just use the value of `align - (buffer.len() % align)` with `buffer.extend(iter::repeat(0).take(n))` here and avoid the while loop.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:1013]: I'd rewrite this whole doc comment to something much simpler without the first paragraph. Maybe just like map BufferUsages to a string.

File: crates/bevy_render/src/render_asset.rs
Comment (JMS55) [inline crates/bevy_render/src/render_asset.rs:171]: I'd like to make these pub in general.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Great docs as always. Looks generally good. If there any bugs or perf overhead from the allocator itself, we can iron them out later. I'm not confident that separate slabs for large buffers are needed, but we can try it out, I don't feel strongly about removing them either. Bevy really needs an example/benchmark for loading and unloading multiple assets in the background during gameplay.

The one thing I'm concerned about is that GpuMesh seems to be in an akward spot now, given that it doesn't actually hold any mesh data anymore. Not sure what we want to do with it now, I'll let @superdump weigh in.

I'd like to see:
* Perf test on WebGL2 to see how much of a hit the allocator logic is vs main
* A bit more detailed migration guide, specifically mentioning the GpuMesh change and how it now lives in MeshAllocator. Please mention the full types, not just the concepts.

**Comment by BD103** [comment]
The original PR was marked for release notes, so I'm going to add it here too!

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:92]: I'm not inclined to indulge SI's attempt to redefine terms that everyone knows :)

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:118]: See https://archive.ph/Z2R8w. tl;dr a value closer to the golden ratio increases the chance that if you reallocate more than once you're able to coalesce older spaces into the new space.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [reply crates/bevy_render/src/mesh/allocator.rs:92]: MiB is base 2 though, instead of base 10, which seems like a better fit for memory sizes to me :)

Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:92]: MB is base 2 for everyone except SI. I don't want to indulge SI's attempt to redefine MB as base 10 for silly reasons.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:125]: This should be MiB since it's 1024 and not 1000 no?

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: I don't like the use of MiB but I changed it anyway.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:125]: I don't like the use of MiB but I changed it anyway.

File: crates/bevy_render/src/render_asset.rs
Comment (pcwalton) [reply crates/bevy_render/src/render_asset.rs]: Done

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: Done

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: Done

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: Removed

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: Fixed

File: crates/bevy_render/src/mesh/mesh/mod.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/mesh/mod.rs]: I renamed it to `RenderMesh` since it's the render world version of a `Mesh`.

File: crates/bevy_render/src/mesh/mesh/mod.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/mesh/mod.rs]: Removed

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (pcwalton) [reply crates/bevy_render/src/batching/gpu_preprocessing.rs]: Reworked all comments here

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: It's because `prepare_assets` clears the contents of the `ExtractedAssets<RenderMesh>` resource after processing it. Because `allocate_and_free_meshes` needs to use that data, it has to run first. (We don't want to merge `allocate_and_free_meshes` with `RenderMesh::prepare_asset` because, among other reasons, that would result in a lot of useless copying when a bunch of meshes load all at the same time and the buffer grows multiple times in a single frame, which happens a lot.)

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: I reworded it.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:1008]: That's wrong in the case in which the buffer is already of the right alignment. But if I wrap it in an if, sure.

Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:1008]: Oh wait, never mind, I have the if already above.

Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: Changed the line to your suggestion.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [inline crates/bevy_render/src/mesh/allocator.rs:812]: You can also use [command_buffer]

File: crates/bevy_render/src/mesh/allocator.rs
Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs:273]: No, we don't, and I don't think it's worth pulling a dependency and bloating Bevy's code size to optimize this. Reallocation is super slow no matter what. The overhead of a couple of CPU side allocations are a drop in the bucket next to allocating the buffer on GPU and copying over.

**Comment by pcwalton** [comment]
I've addressed all the review comments as much as I feel is appropriate. From running the examples on WebGL the performance seems fine. I can't do very in-depth performance analysis there without better diagnostics, and I don't think we should block this PR on better WebGL diagnostics.

**Review by IceSentry — APPROVED** [review]

File: examples/shader/shader_instancing.rs
Comment (tychedelia) [inline examples/shader/shader_instancing.rs:262]: It might be nice to have a quick comment in the example pointing the user to further inspect the docs for `MeshAllocator` to learn more about why this is necessary.

**Review by tychedelia — APPROVED** [review]
Looks good, thanks for the excellent documentation. Two small comments:
- It might be nice to have some logging to warn when a mesh spills over into the large object slab if it's near the boundary.
- We should follow up to rename the other `RenderAsset` e.g `GpuImage` -> `RenderImage  for consistency per the changes here.

**Comment by NthTensor** [comment]
The issue with the bistro interior seems to have been resolved. I can replicate the opaque pass improvements. FPS did not improve in tests of low-end hardware (I am totally gpu bottlenecked on most scenes), but it also doesn't regress which is the main thing I was concerned about.

**Review by NthTensor — APPROVED** [review]

File: examples/shader/shader_instancing.rs
Comment (pcwalton) [reply examples/shader/shader_instancing.rs:263]: It's a borrow check workaround. I wrote a comment above it.

File: crates/bevy_render/src/mesh/allocator.rs
Comment (JMS55) [reply crates/bevy_render/src/mesh/allocator.rs]: Am I correct that this this system runs on a 1-frame delay then?

Comment (pcwalton) [reply crates/bevy_render/src/mesh/allocator.rs]: No. The order in a single frame is this:

1. `extract_render_asset` populates `ExtractedAssets<RenderMesh>` during the extract phase.
2. `allocate_and_free_meshes` examines the `ExtractedAssets<RenderMesh>` to perform the allocations during the asset prep phase.
3. `prepare_asset` processes the `ExtractedAssets<RenderMesh>` again, and clears it out.

So `extract_render_asset` produces the data, `allocate_and_free_meshes` inspects the data, and finally `prepare_asset` consumes the data. This pipeline happens entirely during the frame.

Comment (JMS55) [reply crates/bevy_render/src/mesh/allocator.rs]: Ah ok, thanks for the explanation. This makes sense to me, and mimics how I have it setup in my own code.

**Review by JMS55 — APPROVED** [review]

**Comment by pcwalton** [comment]
The merge fallout is fixed. This is ready to be merged again.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1662 if you'd like to help out.

**Pull request merged**
